File: jAutoInvoice/src/java/net/sf/jautoinvoice/client/modelo/FolhaObra.java
/*
 * FolhaObra.java
 *
 * This file is part of jAutoInvoice, http://sourceforge.net/p/jautoinvoice
 *
 * Copyright (C) 2011 jAutoInvoice
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package net.sf.jautoinvoice.client.modelo;

import java.io.Serializable;
import java.util.ArrayList;

/**
 * 
 * @since 1.0
 */
public final class FolhaObra implements Serializable {

    private Veiculo veiculo;
    private String data;
    private String descricaoAvaria;
    private int kms;
    private Factura factura;
    private ArrayList<LinhaServico> linhas;

    public FolhaObra() {
        linhas = new ArrayList<LinhaServico>();
    }

    public FolhaObra(Veiculo veiculo, String data, String descricaoAvaria, int kms,
            Factura factura, ArrayList<LinhaServico> linhas) {
        this.veiculo = veiculo;
        this.data = data;
        this.descricaoAvaria = descricaoAvaria;
        this.kms = kms;
        this.factura = factura;

        this.linhas = new ArrayList<LinhaServico>();
        if (linhas != null) {
            for (LinhaServico l : linhas) {
                linhas.add(l);
            }
        }
    }

    public void adicionarLinhaServico(LinhaServico linha) {
        if (linhas == null) {
            linhas = new ArrayList<LinhaServico>();
        }

        linhas.add(linha);
    }

    public void removerLinhaServico(LinhaServico linha) {
        if (linhas != null) {
            linhas.remove(linha);
        }
    }

    public String getData() {
        return data;
    }

    public void setData(String data) {
        this.data = data;
    }

    public String getDescricaoAvaria() {
        return descricaoAvaria;
    }

    public void setDescricaoAvaria(String descricaoAvaria) {
        this.descricaoAvaria = descricaoAvaria;
    }

    public Factura getFactura() {
        return factura;
    }

    public void setFactura(Factura factura) {
        this.factura = factura;
    }

    public int getKms() {
        return kms;
    }

    public void setKms(int kms) {
        this.kms = kms;
    }

    public ArrayList<LinhaServico> getLinhas() {
        if (linhas != null) {
            return new ArrayList<LinhaServico>(linhas);
        }

        return null;
    }

    public void setLinhas(ArrayList<LinhaServico> linhas) {
        this.linhas.clear();
        for (LinhaServico l : linhas) {
            linhas.add(l);
        }
    }

    public Veiculo getVeiculo() {
        return veiculo;
    }

    public void setVeiculo(Veiculo veiculo) {
        this.veiculo = veiculo;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final FolhaObra other = (FolhaObra) obj;
        if (this.veiculo != other.veiculo && (this.veiculo == null || !this.veiculo.equals(other.veiculo))) {
            return false;
        }
        if ((this.data == null) ? (other.data != null) : !this.data.equals(other.data)) {
            return false;
        }
        if (this.kms != other.kms) {
            return false;
        }
        if (this.factura != other.factura && (this.factura == null || !this.factura.equals(other.factura))) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        int hash = 3;
        hash = 17 * hash + (this.veiculo != null ? this.veiculo.hashCode() : 0);
        hash = 17 * hash + (this.data != null ? this.data.hashCode() : 0);
        hash = 17 * hash + this.kms;
        hash = 17 * hash + (this.factura != null ? this.factura.hashCode() : 0);
        return hash;
    }
}

File: jai/models/FolhaObra.php
<?php

/* This file is part of jAutoInvoice, a car workshop management software.
 * 
 * Copyright (c) 2012 - 2014, Sérgio Lopes.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @property int $idFolhaObra
 * @property string $data
 * @property string $descricaoAvaria
 * @property integer $kms
 * @property int $activo
 * @property int $idVeiculo
 * @property int $idFuncionario
 * @property string $matricula
 *
 * @property Veiculo $veiculo
 * @property Funcionario $funcionario
 * @property Marcacao $marcacao
 */
class FolhaObra extends CActiveRecord {

    /**
     * @return FolhaObra
     */
    public static function model($className = __CLASS__) {
        return parent::model($className);
    }

    public function tableName() {
        return '{{FolhaObra}}';
    }

    public function rules() {
        return array(
            array('data, kms, idFuncionario, matricula', 'required'),
            array('kms', 'numerical', 'integerOnly' => true),
            array('descricaoAvaria', 'safe'),
            // search
            array('data, descricaoAvaria, idFuncionario, matricula', 'safe', 'on' => 'search'),
        );
    }

    public function relations() {
        return array(
            'veiculo' => array(self::BELONGS_TO, 'Veiculo', 'idVeiculo'),
            'marcacao' => array(self::HAS_ONE, 'Marcacao', 'idMarcacao'),
            'funcionario' => array(self::BELONGS_TO, 'Funcionario', 'idFuncionario'),
            'linhaServicos' => array(self::HAS_MANY, 'LinhaServico', 'idFolhaObra'),
        );
    }

    public function attributeLabels() {
        return array(
            'idFolhaObra' => 'ID',
            'data' => 'Data',
            'descricaoAvaria' => 'Descrição da Avaria',
            'kms' => 'Kms',
            'idVeiculo' => 'Veículo',
            'idFuncionario' => 'Funcionário',
            'matricula' => 'Matrícula'
        );
    }

    /**
     * @return CActiveDataProvider
     */
    public function search() {
        $criteria = new CDbCriteria();

        $criteria->compare('data', $this->data, true);
        $criteria->compare('descricaoAvaria', $this->descricaoAvaria, true);
        $criteria->compare('kms', $this->kms);
        $criteria->compare('matricula', $this->matricula);
        $criteria->compare('idFuncionario', $this->idFuncionario);
        $criteria->compare('activo', 1);

        return new CActiveDataProvider($this, array(
                    'criteria' => $criteria,
                    'sort' => array(
                        'defaultOrder' => 'data DESC',
                        )));
    }

}